### ISSUE TITLE
JSON_parser for json-c is not checking for trailing characters

### ISSUE THREAD

**Issue by anomiex** [facebook/hhvm#5812]
According to the documentation for json-c's json_tokener_parse_ex:
> Trailing characters after the parsed value do not automatically cause an 
> error.  It is up to the caller to decide whether to treat this as an
> error or to handle the additional characters, perhaps by parsing another
> json value starting from that point.

But the implementation of JSON_parser doesn't perform any such check. This leads to input such as `json_decode( 'trueXXXNotReally' )` returning boolean true in HHVM, where PHP returns null with JSON_ERROR_SYNTAX as the last error.

**Comment by Orvid** [comment]
IIRC Assuming json-c is the external json library you can choose to use at compile time, then I believe there might be more issues than this with it, as the internal json parser is used internally in HHVM, but it is disabled when you build with json-c.

**Comment by anomiex** [comment]
Yes, that's it. Due to the "no Evil" clause in the internal library the Debian/Ubuntu packages use json-c. So it'll need to work, hence bug reports.

I should probably figure out how to use github one of these days so I could submit pull requests directly instead of just lobbing bugs.

**Comment by aklapper** [comment]
https://phabricator.wikimedia.org/T107132#1487501 implies that https://github.com/facebook/hhvm/commit/04fd8890b2b9853a51f541f90170ead76da9174b might have fixed this issue?

**Comment by jesseschalken** [comment]
This code now produces the same result on PHP and HHVM

```php
var_dump(json_decode('trueXXXNotReally'));
var_dump(json_last_error());
```
```
NULL
int(4)
```

https://3v4l.org/bla9L

So has this issue been fixed?

**Comment by anomiex** [comment]
3v4l is probably using a build of HHVM that uses the (default?) non-free JSON parser rather than the free json-c parser.

**Comment by lexidor** [comment]
I am going over old issues on this repository, to see which ones apply to the current versions of hhvm.

Support for the json-c parser has been removed. https://github.com/facebook/hhvm/commit/ef919bb81cc2326a24d07bb9b7b7574eced09366
